cube/src/main/java/zone/cogni/asquare/cube/convertor/data2shacl/ShaclGenerator.java

```diff
@@ -325,17 +328,35 @@ private void setShaclDatatype(@Nonnull RdfStoreService rdfStore,
                                 @Nonnull Resource path,
                                 @Nonnull Resource propertyShape) {
     List<String> datatypes = selectUris(rdfStore, "select-datatype.sparql.spel", getTypeAndPropertyParameters(targetClass, path));
-    if (datatypes.size() != 1) {
-      log.warn(getMessage("type '{}' and property '{}' does not have exactly one datatype: {}",
-                          shortenUri(shacl, targetClass), shortenUri(shacl, path), shortenUri(shacl, datatypes)));
+    if (datatypes.isEmpty()) {
+      log.warn(getMessage("type '{}' and property '{}' does not have exactly a datatype",
+                          shortenUri(shacl, targetClass), shortenUri(shacl, path)));
       return;
     }
 
-    Resource datatypeValue = ResourceFactory.createResource(datatypes.get(0));
-    shacl.add(propertyShape, Shacl.datatype, datatypeValue);
+    if (datatypes.size() > 1) {
+      datatypes.forEach(datatype -> {
+        Resource datatypeValue = ResourceFactory.createResource(datatype);
+
+        String orInstanceName = datatypeValue.getLocalName().toLowerCase();
+        Resource singleOrInstance = ResourceFactory.createResource(propertyShape.getURI() + "/" + orInstanceName);
 
-    if (RDF.langString.equals(datatypeValue)) {
-      setLanguageIn(rdfStore, shacl, targetClass, path, propertyShape);
+        shacl.add(propertyShape, Shacl.or, singleOrInstance);
+        shacl.add(singleOrInstance, RDF.type, Shacl.PropertyShape);
+        shacl.add(singleOrInstance, Shacl.datatype, datatypeValue);
+
+        if (datatypeValue.equals(RDF.langString)) {
```

Comment:
will `datatype()` method i sparql return this for "language-strings" ?
did a quick check like this, the xx value is empty:
```SELECT * WHERE {?s skos:prefLabel ?o bind(datatype(?o) as ?xx)} limit 10```

Reply:
(the check was on virtuoso)

Reply:
weird. it works find on Jena... but I guess we must think of supporting Virtuoso too. what do we do? add a TODO?

@domivds